builder/hcloud/step_create_server.go

```diff
@@ -101,3 +111,48 @@ func (s *stepCreateServer) Cleanup(state multistep.StateBag) {
 			"Error destroying server. Please destroy it manually: %s", err))
 	}
 }
+
+func setRescue(ctx context.Context, client *hcloud.Client, server *hcloud.Server, rescue string, sshKeys []*hcloud.SSHKey) error {
+	rescueChanged := false
+	if server.RescueEnabled {
+		rescueChanged = true
+		action, _, err := client.Server.DisableRescue(ctx, server)
+		if err != nil {
+			return err
+		}
+		if err := waitForServerAction(ctx, client, action, server); err != nil {
+			return err
+		}
+	}
+	if rescue != "" {
+		rescueChanged = true
+		res, _, err := client.Server.EnableRescue(ctx, server, hcloud.ServerEnableRescueOpts{
+			Type:    hcloud.ServerRescueType(rescue),
+			SSHKeys: sshKeys,
+		})
+		if err != nil {
+			return err
+		}
+		if err := waitForServerAction(ctx, client, res.Action, server); err != nil {
+			return err
+		}
+	}
+	if rescueChanged {
+		action, _, err := client.Server.Reset(ctx, server)
+		if err != nil {
+			return err
+		}
+		if err := waitForServerAction(ctx, client, action, server); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func waitForServerAction(ctx context.Context, client *hcloud.Client, action *hcloud.Action, server *hcloud.Server) error {
```

Comment:
You should call it something like "waitForAction", because you wait only for server actions in this script.